resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -100,11 +97,53 @@ protected DnsNameResolverContext(DnsNameResolver parent,
         allowedQueries = maxAllowedQueries;
     }
 
-    protected Promise<T> promise() {
-        return promise;
+    void resolve(Promise<T> promise) {
+        resolve(promise, parent.searchDomains().length > 0 && !hostname.endsWith("."));
     }
 
-    void resolve() {
+    private void resolve(Promise<T> promise, boolean trySearchDomain) {
+        if (trySearchDomain) {
+            final Promise<T> original = promise;
+            promise = parent.executor().newPromise();
+            promise.addListener(new FutureListener<T>() {
+                @Override
+                public void operationComplete(Future<T> future) throws Exception {
+                    if (future.isSuccess()) {
+                        original.setSuccess(future.getNow());
+                    } else {
+                        future.addListener(new FutureListener<T>() {
```

Comment:
good catch - yes it can

when you suggested it the first time I was looking at the other promise, not the outer one

> On Jul 7, 2016, at 11:52 PM, Scott Mitchell <notifications@github.com> wrote:
> 
> In resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java <https://github.com/netty/netty/pull/5496#discussion_r69994426>:
> 
> >      }
> >  
> > -    void resolve() {
> > +    private void resolve(Promise<T> promise, boolean trySearchDomain) {
> > +        if (trySearchDomain) {
> > +            final Promise<T> original = promise;
> > +            promise = parent.executor().newPromise();
> > +            promise.addListener(new FutureListener<T>() {
> > +                @Override
> > +                public void operationComplete(Future<T> future) throws Exception {
> > +                    if (future.isSuccess()) {
> > +                        original.setSuccess(future.getNow());
> > +                    } else {
> > +                        future.addListener(new FutureListener<T>() {
> @vietj <https://github.com/vietj> - can't this be reduced to something like this:
> 
>                             final Promise<T> original = promise;
>                             promise = parent.executor().newPromise();
>                             promise.addListener(new FutureListener<T>() {
>                             int count;
>                             @Override
>                             public void operationComplete(Future<T> future) throws Exception {
>                                 if (future.isSuccess()) {
>                                     original.trySuccess(future.getNow());
>                                 } else {
>                                     if (count < parent.searchDomains().length) {
>                                         String searchDomain = parent.searchDomains()[count++];
>                                         Promise<T> nextPromise = parent.executor().newPromise();
>                                         String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
>                                         DnsNameResolverContext<T> nextContext = newResolverContext(parent,
>                                             nextHostname, resolveCache);
>                                         nextContext.internalResolve(nextPromise);
>                                         nextPromise.addListener(this);
>                                     } else {
>                                         original.tryFailure(future.cause());
>                                     }
>                                 }
>                             }
>                         });
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/netty/netty/pull/5496/files/b2995bf1761bcd261a2dee147c5012ef6504100c#r69994426>, or mute the thread <https://github.com/notifications/unsubscribe/AANxilsoozlgSZ_PwlciwI6g0XsWphWbks5qTXUugaJpZM4JEKKu>.
>